### ISSUE TITLE
Missing how-to

### ISSUE THREAD

**Issue by rsre** [RebekkaMa/android-sms-gateway-server#4]
I've seen this in a popular Playbook to deploy a Matrix server. I've checked everywhere for documentation about how to set up this part because I'm really interested, to no avail.

Is there any document where the process is explained?

**Comment by benkuly** [comment]
Currently you need to build and install this app by your own or install it with the apk on the github release page. The app starts a web server on your smartphone which is used by matrix-sms-bridge to send and receive sms.
We are going to document this app within the next few weeks.

**Comment by rsre** [comment]
How can I create or locate the 'Keystore Path' necessary to run the app?

**Comment by benkuly** [comment]
Because the app starts a https-Server it need's a certificate. You can create it with openssl. The matrix-sms-bridge also uses this certificate to validate the host (android-sms-gateway-server). Make sure, that the certificates common-name uses the static ip of the android-sms-gateway-server. Otherwise matrix-sms-bridge could deny the connection.

**Comment by rsre** [comment]
Okey, I tried to select a RSA keypair but it wouldn't let me. I'll try tomorrow to generate a p12 cert and see how it goes. Thank you!

**Comment by tjmccormick** [comment]
Can anyone provide an example of how to create a p12 and set it up?

**Comment by rsre** [comment]
You can generate it using openssl as shown [here](https://stackoverflow.com/questions/21141215/creating-a-p12-file). I haven't tried yet due to lack of free time though.

**Comment by tjmccormick** [comment]
I tried the following:

    openssl genrsa -out key.pem
    openssl req -new -key key.pem -out cert.csr
    openssl x509 -req -in cert.csr -signkey key.pem -out cert.crt
    openssl pkcs12 -export -in cert.crt -inkey key.pem -out cert.p12

On my phone, I select a path to the p12 and enter the password. When I click "Check Certificate Availability" it always says "Error loading certificate:"

**Comment by RebekkaMa** [comment]
This should work: 

```
openssl genrsa -out privkey.pem 4096
openssl req -batch -new -x509 -key privkey.pem -keyform PEM -sha256 -outform PEM -out serverCert.pem
openssl pkcs12 -export -in serverCert.pem -inkey privkey.pem -out yourKeyStoreName.p12 -name yourKeyStoreName -password pass:YourKeystorePassword
```

**Comment by rsre** [comment]
Well, after too much trying I managed to load the p12 certificate.

As a heads up, the file name needs to be the same as created. I try to setup the rest of the process tomorrow.

**Comment by ashemsay** [comment]
I second this request, a succinct how-to would be nice!

**Comment by dzg** [comment]
> This should work:
> 
> ```
> openssl genrsa -out privkey.pem 4096
> openssl req -batch -new -x509 -key privkey.pem -keyform PEM -sha256 -outform PEM -out serverCert.pem
> openssl pkcs12 -export -in serverCert.pem -inkey privkey.pem -out yourKeyStoreName.p12 -name yourKeyStoreName -password pass:YourKeystorePassword
> ```

On macos the 2nd command gives:
```
error, no objects specified in config file
problems making Certificate Request
```

I'm really confused about how to set this up.

The android app doesn't allow to 'start server' ... does it require options first?

How does this work??

**Comment by dzg** [comment]
Anyone? How does this work?

**Comment by rsre** [comment]
I never managed to get it working.

**Comment by lemyoo** [comment]
> Currently you need to build and install this app by your own or install it with the apk on the github release page. The app starts a web server on your smartphone which is used by matrix-sms-bridge to send and receive sms. We are going to document this app within the next few weeks.

Have you been able to do the documentation, please.

**Comment by natecovington** [comment]
Am I following this correctly, I'll need to install a Terminal app on my phone in order to install and use this?

**Comment by benkuly** [comment]
@natecovington You need to build the app yourself or download and install the apk from the release page: https://github.com/RebekkaMa/android-sms-gateway-server/releases

**Comment by natecovington** [comment]
Thank you!   I downloaded the APK from your link onto my phone and it installed:
https://github.com/RebekkaMa/android-sms-gateway-server/releases

Edit: "Start Server" is grayed out - guessing this is a command line only thing

**Comment by johannes-wolf** [comment]
Anyone managed to get it working? I've got the same problem as @natecovington: The start button is grayed out. Also the cert is not working…

**Comment by grojguy** [comment]
> Anyone managed to get it working? I've got the same problem as @natecovington: The start button is grayed out. Also the cert is not working…

No, I have the same problem.  I assume it's because the certificate has to be working first, but I am getting the "error loading certificate" after doing exactly as described [here](https://github.com/RebekkaMa/android-sms-gateway-server/issues/4#issuecomment-818629682). 

@RebekkaMa, or anyone who was successfully getting this set up, could you please post some basic instructions?  

The documentation is very lacking.  Clearly there are some "gotchas" as multiple people here having the same issues.  

Thank you!

**Comment by jpfulton248** [comment]
Though this did indeed "work" to the extent that it allowed me to get the android server running, I still can't actually get the sms bridge to work. I'm using the ansible matrix deploy playbook repo that can be found on github. From tailing the kernlog.log in the matrix-sms-bridge docker container (docker exec -it [containername] /bin/bash) I see certificate path errors. It appears java code (I think running on the bridge side... not the android side) is looking for a file path to a certificate and not finding it. I'm trying to use the same  .p12 keystore file that I'm using on the android side but I can see why that wouldn't logically make sense. I'm going to keep trying and will come back to share and positive progress.

@RebekkaMa 

~~The missing how to for cert creation:~~

~~First, I had to tweak my openssl.cnf. I didn't want to mess with my default so I copied the default, made the necessary modifications (I don't recall exactly but I think just including the alt_names which should be wildcard and naked domain and I'm guessing when it prompts for commondName = FQDN or YOUR name, I typed in mypubliclyaccessibledomain.com) and then specifically pointed at the custom cnf in my commands.~~

```
#opensslsms.cnf
[ req ]
default_bits       = 4096
default_keyfile    = privkey.pem
distinguished_name = req_distinguished_name
req_extensions     = req_ext

[ req_distinguished_name ]
countryName            = Country Name (2 letter code)
stateOrProvinceName    = State or Province Name (full name)
localityName           = Locality Name (eg, city)
organizationName       = Organization Name (eg, company)
commonName             = Common Name (e.g. server FQDN or YOUR name)
emailAddress           = Email Address

[ req_ext ]
subjectAltName = @alt_names

[ alt_names ]
DNS.1 = mypubliclyaccessibledomain.com
DNS.2 = *.mypubliclyaccessibledomain.com

[ v3_ca ]
basicConstraints = critical, CA:TRUE
keyUsage = critical, digitalSignature, keyEncipherment, keyCertSign
```

Once configured, run this:

```
openssl genpkey -algorithm RSA -out privkeysms.pem -aes256 -pkeyopt rsa_keygen_bits:4096
openssl x509 -req -in certreq.csr -signkey privkeysms.pem -out serverCertsms.pem
openssl pkcs12 -export -in serverCertsms.pem -inkey privkeysms.pem -out mypubliclyaccessibledomain.p12 -name mypubliclyaccessibledomain -password pass:MyPassword
```

~~Then copy the 3 resulting files over to your android device (I zipped and emailed to myself). I made a folder somewhere called 'keystore' and put the files there. Then I simply set a username and password in the android app, browsed for my mypubliclyaccessibledomain.p12 file, selected it and then hit check certificate availability. It passed! Server on/off switch no longer greyed out. Upon visiting https://my.phone.private.ip.com:9090, I was prompted to login. I used the username and password that I'd created in the android app and was then presented with:~~

`android sms gateway server (1.0) is running`

**Comment by alexsantee** [comment]
I too struggled with installing it and I'd like to add my findings. First I managed to make it work in my current phone following the author's instructions:

> This should work:
> 
> ```
> openssl genrsa -out privkey.pem 4096
> openssl req -batch -new -x509 -key privkey.pem -keyform PEM -sha256 -outform PEM -out serverCert.pem
> openssl pkcs12 -export -in serverCert.pem -inkey privkey.pem -out yourKeyStoreName.p12 -name yourKeyStoreName -password pass:YourKeystorePassword
> ```

However this did not work at an older phone (Moto G4 with LineageOS Oreo) I had, even using literally the same key file. From `adb logcat` I got this error message with a file that should work and the right password

`MainActivity: isCertificateAvailable: false -> java.io.IOException: exception decrypting data - java.security.NoSuchAlgorithmException: no such algorithm: 1.2.840.113549.1.5.12 for provider BC`

Even tough the toast notification message was the same, I could see if the password was wrong from the logs. Wrong passwords gave a different error message:

`MainActivity: isCertificateAvailable: false -> java.io.IOException: PKCS12 key store mac invalid - wrong password or corrupted file.`

I tried to use slightly different settings for the key, smaller key and DSA algorithm, and using -x509v1 for the certificate to no use. Keystore still tries the same unavailable algorithm.

So my API 27 phone doesn't work because it doesn't have the algorithm in KeyStore, and I didn't find a fix for this

**Comment by Fmstrat** [comment]
I am having the same issue with the start button being grayed out and p12 certificates giving the "Error loading certificate" message. It would be great if @RebekkaMa could chime in here

**Comment by Fmstrat** [comment]
So the app requires an older alg than OpenSSL 3 uses (AES-256/PBKDF2) so I got the app to validate the cert and start with:

```
openssl genrsa -out privkey.pem 4096

openssl req -batch -new -x509 \
  -key privkey.pem \
  -keyform PEM \
  -sha256 \
  -outform PEM \
  -out serverCert.pem \
  -days 36500

openssl pkcs12 -export \
  -legacy \
  -in serverCert.pem \
  -inkey privkey.pem \
  -out server.p12 \
  -name server \
  -password pass:MYPASSWORD
```
